### PR TITLE
Handle 412 error on delete

### PR DESCRIPTION
**Pull request by howdyjessie**
This PR updates how Reference.delete should behave in the event of a 412 error. It implements the logic described in #329 (see the DELETE half of photo-diagram in the thread).

In Reference.delete:
- Do a Reference.read to get current data. If the read fails, retry the read.
- If current data is empty, then the rows aren't in the DB anyway -> resolve Reference.delete.
- If current data and old data match, run Reference.delete again.
- If current data and old data do **not** match, raise 412.
- Retry this 412 error workflow a max of 3 times. If no more tries are left, raise 412.

Also
- The etag property has been moved from Reference to Page
- The Tuple class now has a `page` property which stores the Page that a tuple came from.

**This PR should be merged with informatics-isi-edu/chaise#1006.**

### CODE REVIEW

File: js/reference.js [hunk 1]
```diff
@@ -1164,26 +1171,89 @@ var ERMrest = (function(module) {
...
+                                        mismatchFound = Object.keys(oldTuple).some(findMismatch);
+                                        if (mismatchFound) {
+                                            break;
+                                        }
+                                    }
+                                }
+                                // If old data matches current data, retry the delete w/ updated ETag
+                                if (!mismatchFound && numTriesLeft > 0) {
+                                    numTriesLeft--;
+                                    return ref.delete(tuples);
```
Comment (jrchudy) [inline js/reference.js]: Should this be `currentData`? If not, it doesn't look like you update the `ETag` on `tuples` here before trying to delete again.

File: js/reference.js [hunk 2]
```diff
@@ -1164,26 +1171,89 @@ var ERMrest = (function(module) {
...
                 this._server._http.delete(this.uri, config).then(function deleteReference(response) {
-
                     defer.resolve();
                 }, function error(response) {
-                    var error = module._responseToError(response);
-                    return defer.reject(error);
+                    var status = response.status || response.statusCode;
+                    // If 412 Precondition Failed it means that ETags don't match
+                    if (status == 412) {
+                        var MAX_TRIES = 3, numTriesLeft = MAX_TRIES;
```
Comment (jrchudy) [inline js/reference.js]: I see how the retries are limited by this counter that you decrement. Should there be a delay on this after each retry?

File: js/reference.js (hunk 1 above)
Comment (howdyjessie) [inline js/reference.js]: Yes 🙈 I forgot to re-update this when the etag was moved from Reference. I've updated this to be current data.

File: js/reference.js (hunk 2 above)
Comment (howdyjessie) [inline js/reference.js]: I think that's a good idea. Similar to the retry logic for 500s, I've added a delay that doubles after each retry.

**Pull request merged**